### PR TITLE
first pass at a doc around code and design

### PR DESCRIPTION
**Pull request by maleck13**
@philbrookes @david-martin @aidenkeating @sedroche could you guys give this a read and see does it make sense.

### CODE REVIEW

**Comment by aidenkeating** [comment]
👀

File: docs/server/code.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+## Code
+
+This document will cover the design principals that should guide the server side component for the MCP.
+
+The code development is guided as much as possible by the principals behind Onion architecture or Clean architecture:
```
Comment (aidenkeating) [inline docs/server/code.md]: Are there any videos/resources we could link out to for a more in-depth overview of the architecture?

File: docs/server/code.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+The ui for the MCP is embedded into the OpenShift UI, we are able to reuse the bearer token created when the user first logs into OpenShift.
+From the UI, this token is sent with each request to the server. Using this token the access middleware first checks if it needs to check access to the currently
+requested URL, if it does, it makes a request to the user read endpoint in OpenShift. If we get a valid response, then we know the user is a 
+valid user for this instance of OpenShift and we can proceed.
+
+**Authorization**
+
+Once a user is authenticated, we use their bearer token in order to interact with the kubernetes and openshift API. In this way we offload 
+authorisation to the cluster. However is some places it is useful to check access to a particular resource within a given namespace. To do this we 
+do a localaccessreview.
```
Comment (philbrookes) [inline docs/server/code.md]: not sure if it matters, but I've actually ended up using a `localsubjectaccessreviews`

File: docs/server/code.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+In the MCP server we also have available a service account that is granted the edit role on the namespace where the server is deployed.
+This service account is used when preforming tasks on behalf of a mobile client: for instance, a mobile client will request the sdk service 
+configuration from the server. We cannot pass a bearer token from a mobile client. In this case we authenticate the mobile client by its
+mobile clientID and its generated API Key we then use the service account token to configure our clients and act against the namespace.
+
+
+### The Data Layer
+
+MCP does not have a direct dependency on a database of its own. Instead, it levarages data structures provided by the Kubernetes API
```
Comment (philbrookes) [inline docs/server/code.md]: leverages*

File: docs/server/code.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+in order to store data. So we use objects such as [configmaps](https://kubernetes-v1-4.github.io/docs/user-guide/configmap/) and [secrets](https://kubernetes-v1-4.github.io/docs/user-guide/secrets/) in order to store data that is used to represent our mobile
+domain objects. For example a mobile app is backed by a configmap. However the data layer (contained in ```pkg/data```) is the only layer that knows there is a configmap,
+every other layer uses an instance of ```mobile.App```. The data layer is responsible for taking a storing domain objects and returning domain objects from whatever backing store is used to store them.
```
Comment (philbrookes) [inline docs/server/code.md]: `taking a storing`? Is that `taking and storing`?

File: docs/server/code.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+
+### Adding and testing a new web handler
+
+The web layer is responsible for
+accepting requests, parsing the data sent in those requests into the correct domain objects, and handling other request response based actions, this layer should not contain business logic.
+
+Web handlers live under the ```pkg/web``` directory as do the tests that test the handlers. As we base our authentication and authorization from the token passed with
+each request, these handlers are also required to configure the kubernetes clients with the correct token to ensure we are acting as the requesting user.
+These configured dependencies implement interfaces required by the service business logic. When a request comes in, a cofigured instance is passed through
```
Comment (philbrookes) [inline docs/server/code.md]: `cofigured` > `configured`

File: docs/server/code.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+**Authentication**
+
+The ui for the MCP is embedded into the OpenShift UI, we are able to reuse the bearer token created when the user first logs into OpenShift.
+From the UI, this token is sent with each request to the server. Using this token the access middleware first checks if it needs to check access to the currently
+requested URL, if it does, it makes a request to the user read endpoint in OpenShift. If we get a valid response, then we know the user is a 
+valid user for this instance of OpenShift and we can proceed.
+
+**Authorization**
+
+Once a user is authenticated, we use their bearer token in order to interact with the kubernetes and openshift API. In this way we offload
```
Comment (aidenkeating) [inline docs/server/code.md]: openshift -> OpenShift will keep the document consistent

File: docs/server/code.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+**Authorization**
+
+Once a user is authenticated, we use their bearer token in order to interact with the kubernetes and openshift API. In this way we offload 
+authorisation to the cluster. However is some places it is useful to check access to a particular resource within a given namespace. To do this we 
+do a localaccessreview. 
+
+**Service Account**
+
+In the MCP server we also have available a service account that is granted the edit role on the namespace where the server is deployed.
+This service account is used when preforming tasks on behalf of a mobile client: for instance, a mobile client will request the sdk service
```
Comment (aidenkeating) [inline docs/server/code.md]: preforming -> performing

**Comment by david-martin** [comment]
:eyes:

File: docs/server/code.md (hunk 1 above)
Comment (david-martin) [reply docs/server/code.md]: +1 on this. Linking to existing stuff would be useful

File: docs/server/code.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+- Dependency Inversion: dependencies should always point in to the layers below and never back out.
+- Dependency injection: defining interfaces that can be swapped out at runtime for different concrete implementations
+- The domain model should always be at the very center and depend on nothing external.
+
+### Layers: 
+
+#### Domain Model
+
+This is the core layer and should depend on none of the other pkgs that make up the application. Our domain model is
+contained in the mobile package as mobile is our core domain the actual model definitions are located in ``` pkg/mobile/types.go```
```
Comment (david-martin) [inline docs/server/code.md]: Might read better with a stop here "is our core domain. The actual"

File: docs/server/code.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+is not meant to make up part of the core domain model.
+The core application interfaces are also located in the core mobile package under ```interfaces.go``` these interfaces, define how the core mobile 
+layer expects to be able to interact with external services such as database or web service in order to fulfill its business logic requirements. It is 
+the responsibility of these external layers to fulfill this interface.
+
+
+#### Services (Business logic)
+
+Our services or business logic are contained within the sub packages of the core mobile package. These packages can depend on the core mobile
+package and domain model, but not the other way around, they should also not depend on any database specific implementation or know anything about http.
```
Comment (david-martin) [inline docs/server/code.md]: re: "not depend on any database specific implementation or know anything about http",
this is not the case for keycloak & fh-sync-server in the metrics pacakge.
They uses a requester, which deals with requests & responses. And there is setting of the 'Authorization' header in the keycloak one. Should we abstract that away further or move those functions out of those files completely?

File: docs/server/code.md (hunk 2 above)
Comment (david-martin) [reply docs/server/code.md]: any docs or code we can link to that gives more info on what this is?

File: docs/server/code.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+The core application interfaces are also located in the core mobile package under ```interfaces.go``` these interfaces, define how the core mobile 
+layer expects to be able to interact with external services such as database or web service in order to fulfill its business logic requirements. It is 
+the responsibility of these external layers to fulfill this interface.
+
+
+#### Services (Business logic)
+
+Our services or business logic are contained within the sub packages of the core mobile package. These packages can depend on the core mobile
+package and domain model, but not the other way around, they should also not depend on any database specific implementation or know anything about http. 
+They should only contain the mobile specific business logic and consume interfaces for interacting with things like a database or external services. They consume the and act on domain models and perform logic using these models.
```
Comment (sedroche) [inline docs/server/code.md]: "They consume the and act on domain models..." Should this be "They consume the *interface* and act on domain models..."

**Comment by maleck13** [comment]
updated as per reviews

**Pull request merged**